modules/README.md

```diff
@@ -1,58 +1,164 @@
+# Module requirements
 
-# PXP agent architecture
+### Metadata
 
-map of name:module, possibly factory style from peg
+The `metadata` of a module is a JSON object that contains a number of JSON
+schemas for specifying configuration options and actions. It contains:
 
-Module abstraction
+ - **configuration**: (optional) schema that describes the module configuration format;
+ - **actions**: an array where each item is an object that describes an action implemented by the module (please, refer to the below schema).
 
-    class Module {
-        std::string name;
-        bool invoke(const CthunMessage& request, CthunMessage& response);
-        Schema actions;
+The `metadata` schema is:
+
+    {
+        "type" : "object",
+        "description" : "Metadata format",
+        "properties" : {
+            "actions" : {
+                "type" : "array",
+                "items" : {
+                    "type" : "object",
+                    "description" : "Describes an action",
+                    "properties" : {
+                        "name" : {
+                            "type" : "string",
+                            "description" : "Name of the action",
+                        },
+                        "input" : {
+                            "type" : "object",
+                            "description" : "Schema for the PXP request arguments",
+                        },
+                        "results" : {
+                            "type" : "object",
+                            "description" : "Schema for the action results",
+                        },
+                    },
+                    "required" : ["description", name", "input", "results"],
```

Comment:
Right, I thought it may get confusing.
This JSON object is meant to be a schema for `metadata`, os it's basically a schema containing other schemas ("configuration and, for each action, "input" and "results" are all JSON schemas).
So, on one hand, we're not specifying the structure of each required schema; we're just stating that we need those entries. On the other, we're using a schema for the whole metadata structure - from http://json-schema.org/example1.html:
'''
the title and description keywords are descriptive only, in that they do not add constraints to the data being validated. 
'''

It may be the best way to specify the `metadata` structure, but in this case I believe that using `description` is ok.